src/core/TESTING.md

```diff
@@ -59,11 +59,294 @@ test('my test', async () => {
+        .expect(400);
+
+      expect(response.body).to.have.property('message');
+      expect(response.body.message).to.contain('must have a maximum length of [100]');
+    });
+  });
+```
+- the main logic of the plugin
+```ts
+export default function({ getService }: FtrProviderContext) {
+  const supertest = getService('supertest');
+  describe('myPlugin', () => {
+    it('stores something', async () => {
+      const response = await supertest
+        .post('/myPlugin/something')
+        .set('content-type', 'application/json')
+        .send({ text: 'aaa' })
+        .expect(200);
+
+      expect(response.body).to.have.property('id');
+      expect(response.body.id).to.be.a('string');
+    });
+
+    it('retrieves something', async () => {
+      const { body } = await supertest
+        .post('/myPlugin/something')
+        .set('content-type', 'application/json')
+        .send({ text: 'bbb' })
+        .expect(200);
+
+      const response = await supertest.get(`/myPlugin/something/${body.id}`).expect(200);
+      expect(response.text).be('bbb');
+    });
+
+    it('returns NotFound error when cannot find a resource', async () => {
+      await supertest
+        .get('/myPlugin/something/missing')
+        .expect(404, 'Saved object [myPlugin-type/missing] not found');
+    });
+  });
+```
 
-#### HTTP Routes
+##### TestUtils
+It can be utilized if your plugin doesn't interact with the elasticsearch server. Runs tests against real Kibana server instance.
+Pros:
+- runs the real Kibana instance
+- tests cross-plugin integration
+- emulates a real user interaction with the HTTP resources
+Cons:
+- faster than FTR because it doesn't run Elasticsearch instance, but still slow
+- hard to debug
+- doesn't cover Kibana CLI logic
+
+###### Example
+Given `myPlugin` plugin with an endpoint to format user-provided text:
+```ts
+router.get(
+  {
+    path: '/myPlugin/formatter',
+    validate: {
+      query: schema.object({
+        text: schema.string({ maxLength: 100 }),
+      }),
+    },
+  },
+  async (context, request, response) => {
+    try {
+      const sanitizeString = deps.sanitizer.sanitize(request.query.text);
+      const formattedText = format(sanitizeString);
+      return response.ok({ body: formattedText });
+    } catch(error) {
+      if (error instanceof MisformedTextError) {
+        return response.badRequest({ body: error.message })
+      }
+    }
+  }
+);
+```
+To have access to Kibana TestUtils, you should create `integration_tests` folder and import `test_utils` within a test file:
+```ts
+// src/plugins/my_plugin/server/integration_tests/formatter.test.ts
+import * as kbnTestServer from 'src/test_utils/kbn_server';
+
+describe('myPlugin', () => {
+  describe('GET /myPlugin/formatter', () => {
+    let root: ReturnType<typeof kbnTestServer.createRoot>;
+    beforeAll(async () => {
+      root = kbnTestServer.createRoot();
+      await root.setup();
+      await root.start();
```

Comment:
Having a new server for every test might be quite slow. Since HTTP requests should be stateless I don't think tests would normally interfere with each other if they all shared the same server instance. Perhaps we can add a comment like "Unless your tests are likely to interfere with each other, consider sharing a single server instance between all tests to improve performance"

Reply:
👍 I can add it prior to the `beforeAll` hook.